specs/experimental/gov-delegation.md

```diff
@@ -0,0 +1,406 @@
+    - [`numCheckpoints`](#numcheckpoints)
+    - [`delegates`](#delegates)
+    - [`getVotes`](#getvotes)
+    - [`getPastVotes`](#getpastvotes)
+    - [`getPastTotalSupply`](#getpasttotalsupply)
+    - [`migrated`](#migrated)
+    - [`delegations`](#delegations)
+  - [Events](#events)
+    - [`DelegationCreated`](#delegationcreated)
+    - [`DelegateVotesChanged`](#delegatevoteschanged)
+- [Storage](#storage)
+- [Constants](#constants)
+- [Types](#types)
+  - [`Delegation`](#delegation)
+  - [`AllowanceType`](#allowancetype)
+- [Backwards Compatibility](#backwards-compatibility)
+- [Migration](#migration)
+- [User Flow](#user-flow)
+  - [Partial Delegations](#partial-delegations)
+  - [Absolute & Relative Delegations](#absolute--relative-delegations)
+  - [Differences](#differences)
+- [Delegation Validation](#delegation-validation)
+- [Security Considerations](#security-considerations)
+  - [Dependence on GovernanceDelegation](#dependence-on-governancedelegation)
+  - [Connection with GovernanceToken](#connection-with-governancetoken)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+| Constant | Value                                        |
+|----------|----------------------------------------------|
+| Address  | `0x4200000000000000000000000000000000000043` |
+
+The `GovernanceDelegation` contract implements advanced delegation for the [`GovernanceToken`](gov-token.md).
+Advanced delegation allows for partial delegations of voting power using absolute or relative amounts. Absolute
+delegations are fixed and denominated in the `GovernanceToken`'s decimals, while relative delegations are denominated
+in basis points.
+
+By using transfer hooks in the `GovernanceToken`, the `GovernanceDelegation` contract can update the delegation
+state and checkpoints after token transfers. This mechanism ensures that any delegations are correctly applied
+and that the voting power of users is accurately reflected as token balances change.
+
+## Interface
+
+### Core Functions
+
+#### `delegate`
+
+Delegates voting power to another address (delegatee) with the given delegation parameters. This function
+is intended to be called by users that require advanced delegation of the `GovernanceToken`, and overwrites
+any existing delegations of the delegator.
+
+```solidity
+function delegate(Delegation _delegation) external
+```
+
+This function MUST enforce the migration logic, as specified in the [Migration](#migration) section, for `msg.sender`
+and the `_delegatee` address in the delegation. Afterwards, the function MUST call the [`_delegate`](#_delegate) function
+with the `msg.sender` and `_newDelegations` parameters.
+
+At the end, the `delegate` function MUST emit a `DelegationCreated` event with the given function parameters.
+
+#### `delegateFromToken`
+
+Delegates 100% of the token voting power of an address (delegator) to another address (delegatee), mimicking the behavior
```

Comment:
Minor comment, work explicitly calling out that this overwrites all existing delegations.